### DISCUSSION TITLE
It's about `type`

### DISCUSSION THREAD

**Discussion by thorwhalen** [i2mint/i2mint#3] [Enhancement]
This discussion is about [typing](https://en.wikipedia.org/wiki/Type_system).

In python, unlike most languages, we have the choice. We can type everything... or nothing... and everywhere in between. 

So where should we place ourselves?

I'd like us to come up with:
* Some rules of thumb to know when to type and when not to
* How to type
* A set of ready-to-use types for our particular domain(s)
  * including, associated with the type, a collection of commonly associated variable names (which is what I started in [atypes](https://github.com/otosense/atypes))

## A few typing tools to consider

We favor builtins first -- so pretty much just
* [typing](https://docs.python.org/3/library/typing.html)
* [types](https://docs.python.org/3/library/types.html)
* [collections.abc](https://docs.python.org/3/library/collections.abc.html)

But for some situations, we should really reach out to third-party tools
* [mypy](https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html#cheat-sheet-py3), which we could use in our IDEs and CI
* [pydantic](https://docs.pydantic.dev/latest/), the most widely used data validation library, with a rich ecosystem that we should know more about. We definitely should add a layer of `pydantic` over stuff we will use in apps (so useful to have tooling for this in [front](https://github.com/i2mint/front) and/or [extrude](https://github.com/i2mint/extrude), and perhaps optionally (in an "optional import conditioned" module?) [atypes](https://github.com/otosense/atypes).

## Advantages of Using a Type System:

I'll let Geppetto say a few words about that...

- **Error Detection:**
   - Helps in catching errors at compile-time rather than at runtime, which can prevent many common bugs related to incorrect data types.
- **Readability and Maintainability:**
   - Provides self-documentation by making the code more expressive, allowing developers to understand the structure and behavior of the software better.
- **Optimization:**
   - Enables compilers to perform optimizations that can make the code run faster and use resources more efficiently.
- **Refactoring Safety:**
   - Makes refactoring and modifying code safer and more predictable by catching incompatible changes.
- **Enhanced Tooling:**
   - Supports better development tools, providing features like auto-completion, type-checking, and refactoring tools, which can improve developer productivity.
- **Robustness:**
   - Contributes to the robustness and correctness of the software by ensuring that the values in the program adhere to the expected types.

### Disadvantages of Using a Type System:

- **Learning Curve:**
   - Requires time to learn and can be complex, especially for newcomers to programming or a specific language.
- **Verbosity:**
   - Can lead to more verbose code due to the need to declare types, which can be seen as additional overhead.
- **Development Speed:**
   - May slow down the initial development speed due to the added complexity and the need to define types explicitly.
- **Flexibility:**
   - Can restrict flexibility and expressiveness in coding due to rigid type constraints, especially in cases where dynamic typing could be beneficial.
- **Complexity:**
   - Introduces additional complexity to the language and development process, which might not be necessary for smaller or less complex projects.

**Comment by thorwhalen** [comment]
## Use `TypeAlias`

Should use [TypeAlias](https://docs.python.org/3/library/typing.html#typing.TypeAlias) once we fully move to 3.10. 

We could start using it by conditioning the import (falling back to `typing_extensions`, which contains it for earlier version of python.

### Problem

The following works:

```
from typing import Dict
RenderKeyEdits = Dict[str, dict]
```

So why do the following make my linter complain?

```
from typing import Dict, NewType
RenderKeyEdits = NewType('RenderKeyEdits', Dict[str, dict])
```

### Solution

The `NewType` function in the typing module is used to create a new type that is distinct from an existing type. When you define a new type using `NewType`, you create a new class that is a subclass of the original type. However, the type checker doesn't automatically treat the new type as a subtype of the original type.

In your case, you are defining a new type `RenderKeyEdits` using `NewType`, which is a subclass of `Dict[str, dict]`. However, when you use `RenderKeyEdits` in a type hint, `mypy` doesn't recognize it as a subtype of Dict[str, dict]. To make `mypy` recognize the new type as a subtype of the original type, you can use the `TypeAlias` function from the `typing_extensions` module:

```python
from typing import Dict
from typing_extensions import TypeAlias, NewType

RenderKeyEdits = NewType('RenderKeyEdits', Dict[str, dict])
RenderKeyEditsAlias = TypeAlias[Dict[str, dict], RenderKeyEdits]
```

In this example, `RenderKeyEditsAlias` is defined as a type alias that maps the original `Dict[str, dict]` type to the `RenderKeyEdits` type created using `NewType`. You can use `RenderKeyEditsAlias` in your type hints to specify that the value should be of type `Dict[str, dict]` or `RenderKeyEdits`.

**Comment by thorwhalen** [comment]
## Granularity of collection types

I saw this code:

```python
import os
from config2py import get_config as _get_config
from typing import Union, Optional, Collection, Callable, Mapping, Any

ConfigSource = Union[Callable, Mapping[str, Any]]

def get_config(
        key: str,
        sources: Optional[Collection[ConfigSource]] = None
) -> str:
    sources = sources or [os.environ]
    return _get_config(key, sources)
```

### My comment

(...) Perhaps it's not `Collection[ConfigSource]` that should be used here but `Iterable[ConfigSource]` since we don’t really need to support `len` (`__len__` method) or `in` operator (`__contains__` method). See [collections.abc](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) for definitions. 

### The reply

Regarding the choice between Iterable, `Collection`, or even `Sequence`, it's a dilemma I am regularly faced with. To decide on what would make the most sense, the question I ask myself is not whether the object should support `len` or reverse calls, but rather if it would make sense if the object could generate unordered or (potentially) infinite items.

Here, the order of the config sources doesn't (shouldn't) matter, but it wouldn't make sense to have an object that could generate an infinite amount of sources. Hence my decision to put `Collection` as a type hint, to indicate this to the user, even though there's probably no call to `len` in the function implementation.
I guess that could potentially lead to warning or errors when using a type checker, if you send an custom object with no `__len__` method implemented. But, on the other hand, wouldn't it be a bit risky to send to this function an object that could potentially generate an infinite amount of sources?

### My reply on that

I feel your pain. 

It's absolutely a dilemma I often have. 
In fact, I usually hit a wall of expressivity quite fast when doing types in python. 
Perhaps I just don't know the tricks yet, or perhaps the choice the python gods made for its type system were just biased in a different direction. 

I for one, would like more granularity and operability in typing. 
I'd like it to be a lot more behavior and role oriented. 
I feel like, even if many decades separate us to the days where the number of bytes reserved for some type was an actual concern, we still live under the shadow of that legacy. 

#### About `typing.Protocol`

The only builtin tool that satisfies my itch for this is the [typing.Protocol](https://docs.python.org/3/library/typing.html#typing.Protocol) (see also [mypy's page for protocols](https://mypy.readthedocs.io/en/stable/protocols.html) and the [pep544](https://peps.python.org/pep-0544/).

For example, it's what I use because I often want to just express that an object `obj` is [Gettable](https://github.com/otosense/atypes/blob/3d104f01c78d6c717e6bc3f2bb5f965051422cc5/atypes/protocols.py#L9), that is, has a `__getitem__` so that it can one can do `obj[key]` on it, or, adding a stronger condition on that, that it's `[Sliceable](https://github.com/otosense/atypes/blob/3d104f01c78d6c717e6bc3f2bb5f965051422cc5/atypes/protocols.py#L36)` which expresses the fact that you can do `obj[start:end]`. 
Both of those are in in the [atypes.protocols](https://github.com/otosense/atypes/blob/master/atypes/protocols.py) module, and vendored (a guilt-minimizing way of saying "copy/pasted from") in many other libraries of ours. 

That said, my dissatisfaction here is that I don't get the functionality I want out of `Protocol`. 
I can express details of the methods and their signatures, but neither static typing nor dynamic `isinstance` checks (when using `typing.runtime_checkable` on the protocol) seem to even look at the signatures. 

#### About finiteness and order

I like the finiteness concern. I'm not sure having a `__len__` completely solves it, but it's definitely the best I can think of. There is no builtin "is finite" type or special method as far as I know. I do believe it's a good one to have though. Maybe [pydantic](https://docs.pydantic.dev/latest/usage/types/types/) has some type for that?

Now, on the particular subject of `get_config`, I don't see why finiteness is a functional condition on `sources`. You could, potentially, have an infinite (probably a loop on a finite) iterable of sources. The function would then search a config is found. Unbounded waiting is not uncommon (it happens almost anytime computers reach out and wait for user input), and anyway, we can never be sure that function will ever return (that's the [halting problem](https://en.wikipedia.org/wiki/Halting_problem)).

So my conclusion is: In this particular case, we definitely don't need a `__contains__`, and just argued we don't need a `__len__`. All we need is to be able to loop through the `sources`. 

On the other hand, the order consideration is a good one, and very interesting to think about. 
Because the function doesn't depend on that characteristic to **function**, but does depend on it to **function as expected**. If you give it a set of sources, you have no assurance of where the config will come from or even if you'll get it from the same place every time and every system. 

It's a weird one. Because what you want is not "the iterable is ordered", you can only say "the iterable is ordered as the user wants it to be". On the other hand we **do** have a proxy for "the iterable HAS an order" -- namely that it has an integer index. So, if we can express, as a type, the fact that one can do a `obj[index]` on `obj` where `idx` is an integer, then we have something. 

Again, you can do that with a `Protocol` -- see my `Getter` and `Sliceable` I mentioned above. 
But, as mentioned above, that level of detail on the signature of the methods seems to be ignored.

I'd like to know more about why, and if there's a possibility to address that!

### Other related stuff

* [From annotated functions to meshes](https://github.com/i2mint/meshed/discussions/55)

### Appendix


Now, it's not the point here, but perhaps a better solution would have been to do, simply:

```python
from functools import partial
get_config = partial(_get_config, sources=(os.getenv,))
```

Here, we use `os.getenv` instead of os.environ because that latter would show all the values (some secret) of the environment variables in the signature, and, depending on the way it's serialized, could even end up serializing them (and sharing them with others). 

Also, we're using a tuple, instead of a list, in the defaults, because it's a better practice to put immutable defaults in functions.

**Reply by charlie-dufort** [reply]
> [...] [typing.Protocol](https://docs.python.org/3/library/typing.html#typing.Protocol) (see also [mypy's page for protocols](https://mypy.readthedocs.io/en/stable/protocols.html) and the [pep544](https://peps.python.org/pep-0544/).  

I'll add that to the infinite list of stuff I'd like to read about Python..
